apps/prairielearn/src/tests/lti13.test.ts

```diff
@@ -0,0 +1,248 @@
+
+    const savePlatformOptionsButton = ltiInstanceResponse.$(
+      'button:contains(Save platform options)',
+    );
+    const platformOptionsForm = savePlatformOptionsButton.closest('form');
+
+    // Update the platform options.
+    const updatePlatformOptionsResponse = await fetchCheerio(
+      `${siteUrl}/pl/institution/1/admin/lti13/1`,
+      {
+        method: 'POST',
+        body: new URLSearchParams({
+          __csrf_token: platformOptionsForm.find('input[name=__csrf_token]').val() as string,
+          __action: platformOptionsForm.find('input[name=__action]').val() as string,
+          platform: 'Unknown',
+          issuer_params: JSON.stringify({
+            issuer: `http://localhost:${oidcProviderPort}`,
+            authorization_endpoint: `http://localhost:${oidcProviderPort}/auth`,
+            jwks_uri: `http://localhost:${oidcProviderPort}/jwks`,
+          }),
+          custom_fields: '{}',
+          client_id: CLIENT_ID,
+        }),
+      },
+    );
+    assert.equal(updatePlatformOptionsResponse.status, 200);
+
+    const addKeyButton = updatePlatformOptionsResponse.$('button:contains(Add key to keystore)');
+    const keystoreForm = addKeyButton.closest('form');
+
+    // Create a key
+    const createKeyResponse = await fetchCheerio(`${siteUrl}/pl/institution/1/admin/lti13/1`, {
+      method: 'POST',
+      body: new URLSearchParams({
+        __csrf_token: keystoreForm.find('input[name=__csrf_token]').val() as string,
+        __action: addKeyButton.attr('value') as string,
+      }),
+    });
+    assert.equal(createKeyResponse.status, 200);
+  });
+
+  step('enable LTI 1.3 as an authentication provider', async () => {
+    const ssoResponse = await fetchCheerio(`${siteUrl}/pl/institution/1/admin/sso`);
+    assert.equal(ssoResponse.status, 200);
+
+    const saveButton = ssoResponse.$('button:contains(Save)');
+    const form = saveButton.closest('form');
+    const lti13Label = form.find('label:contains(LTI 1.3)');
+    const lti13Input = lti13Label.closest('div').find('input');
+
+    const enableLtiResponse = await fetchCheerio(`${siteUrl}/pl/institution/1/admin/sso`, {
+      method: 'POST',
+      body: new URLSearchParams({
+        __csrf_token: form.find('input[name=__csrf_token]').val() as string,
+        __action: saveButton.attr('value') as string,
+        enabled_authn_provider_ids: lti13Input.attr('value') as string,
+      }),
+    });
+    assert.equal(enableLtiResponse.status, 200);
+  });
+
+  step('perform login', async () => {
+    // `openid-client` relies on the session to store state, so we need to use
+    // a cookie-aware version of fetch.
+    const fetchWithCookies = fetchCookie(fetchCheerio);
+
+    const startLoginResponse = await fetchWithCookies(`${siteUrl}/pl/lti13_instance/1/auth/login`, {
+      method: 'POST',
+      body: new URLSearchParams({
+        iss: siteUrl,
+        // TODO: what should this be?
+        login_hint: 'login_hint',
```

Comment:
Per the [IMS Global Security Framework](https://www.imsglobal.org/spec/security/v1p0/), `login_hint` is mandatory and is defined like this:

> Hint to the Authorization Server about the login identifier the End-User might use to log in. The permitted values will be defined in the host specification.

I wish the spec said more about this. In your example login request/response flow, `login_hint` corresponds to `"https://purl.imsglobal.org/spec/lti/claim/lti11_legacy_user_id"` and `"https://purl.imsglobal.org/spec/lti/claim/lti1p1"["user_id"]`. Reading Canvas's docs (https://canvas.instructure.com/doc/api/file.lti_dev_key_config.html), they say that this value just has to be passed *back* to Canvas unchanged. Is there anything else you know about this value that could be relevant?

I'm thinking I should test that this value is present in the redirect to the provider, for completeness.